src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -227,8 +361,502 @@ sealed abstract class AST(pos: Position, subexprs: Array[AST] = Array.empty) {
+      ec <- ec();
+      compiledCode = cc.runWithDelayedValues(name +: bindings.map(_._1).toSeq, ec);
+      f <- invokePrimitive1(((vs: Array[AnyRef]) => (x: AnyRef) => compiledCode((x +: vs).to[mutable.ArrayBuffer])).asInstanceOf[AnyRef => AnyRef])(cvalues)
+    ) yield f.asInstanceOf[Code[AnyRef => T]]
+  }
+
+  def oOrNull[T >: Null](x: Code[Option[T]])(implicit tct: ClassTag[T]): CM[Code[T]] =
+    CM.memoize(x).map { case (st, opt) => Code(st, opt.invoke[Boolean]("isEmpty").mux(Code._null[T], opt.invoke[T]("get"))) }
+
+  def mapO[T, U >: Null](x: Code[Option[T]], f: Code[T] => Code[U])(implicit tct: ClassTag[T]): CM[Code[U]] =
+    CM.memoize(x).map { case (st, opt) => Code(st, opt.invoke[Boolean]("isEmpty").mux(Code._null[U], f(opt.invoke[T]("get")))) }
+
+  def mapOM[T, U >: Null](x: Code[Option[T]], f: Code[T] => CM[Code[U]])(implicit tct: ClassTag[T], uti: TypeInfo[U]): CM[Code[U]] = for (
+    (st, opt) <- CM.memoize(x);
+    out <- f(opt.invoke[T]("get"))
+  ) yield
+    Code(st, opt.invoke[Boolean]("isEmpty").mux(Code._null[U], out))
+
+  def mapA[T, U](ca: Code[Array[T]], f: Code[T] => CM[Code[U]])(implicit tct: ClassTag[T], tti: TypeInfo[T], atct: ClassTag[Array[T]], uct: ClassTag[U], uti: TypeInfo[U], auti: TypeInfo[Array[U]]): CM[Code[IndexedSeq[U]]] = for (
+    (sta, a) <- CM.memoize(Code.checkcast[Array[T]](ca))(arrayInfo[T]);
+    (stn, n) <- CM.memoize(toCodeArray(a).length());
+    (stb, b) <- CM.memoize(Code.newArray[U](n));
+    i <- CM.newLocal[Int];
+    codeAr = toCodeArray(a);
+    newElement <- f(codeAr(i))
+  ) yield Code(
+    i.store(0),
+    sta, stn, stb,
+    Code.whileLoop(i < n,
+      Code(b.update(i, newElement), i.store(i + 1))
+    ),
+    CompilationHelp.arrayToWrappedArray(b)
+  )
+
+  def mapIS[T, U](ca: Code[IndexedSeq[T]], f: Code[T] => CM[Code[U]])(implicit tct: ClassTag[T], tti: TypeInfo[T], atct: ClassTag[Array[T]], uct: ClassTag[U], uti: TypeInfo[U], auti: TypeInfo[Array[U]]): CM[Code[IndexedSeq[U]]] = for (
+    (sta, _a) <- CM.memoize(ca);
+    a = Code.checkcast[scala.collection.SeqLike[T, IndexedSeq[T]]](_a);
+    (stn, n) <- CM.memoize(Invokeable.lookupMethod[scala.collection.SeqLike[T, IndexedSeq[T]], Int]("size", Array()).invoke(a, Array()));
+    (stb, b) <- CM.memoize(Code.newArray[U](n));
+    i <- CM.newLocal[Int];
+    oldElement = Invokeable.lookupMethod[scala.collection.SeqLike[T, IndexedSeq[T]], T]("apply", Array(implicitly[ClassTag[Int]].runtimeClass)).invoke((a), Array(i));
+    newElement <- f(oldElement)
+  ) yield Code(
+    i.store(0),
+    sta, stn, stb,
+    Code.whileLoop(i < n,
+      Code(b.update(i, newElement), i.store(i + 1))
+    ),
+    CompilationHelp.arrayToWrappedArray(b)
+  )
+
+
+  def funreg(name: String, args: Seq[AST], argTypes: Seq[Type]): CM[Code[AnyRef]] = {
+    val m = FunctionRegistry.lookup(name, MethodType(argTypes: _*))
+      .valueOr(x => fatal(x.message))
+    m match {
+      case aggregator: Arity0Aggregator[_, _] =>
```

Comment:
These three are pretty ugly, IMO, but I don't have a better answer yet. I think I need to change the Aggregator interface a bit.